cpp/src_prims/sparse/coo.h

```diff
@@ -51,113 +53,130 @@ namespace Sparse {
  * while the original caller still maintains ownership of the underlying memory.
  *
  * @tparam T: the type of the value array.
+ * @tparam Index_Type: the type of index array
  *
  */
-template <typename T>
+template <typename T, typename Index_Type = int>
 class COO {
  protected:
-  bool owner;
+  device_buffer<Index_Type> rows;
+  device_buffer<Index_Type> cols;
+  device_buffer<T> vals;
 
  public:
-  int *rows;
-  int *cols;
-  T *vals;
-  int nnz;
-  int n_rows;
-  int n_cols;
-  bool device;
+  Index_Type nnz;
+  Index_Type n_rows;
+  Index_Type n_cols;
 
   /**
-        * @param device: are the underlying arrays going to be on device?
-        */
-  COO(bool device = true)
-    : rows(nullptr),
-      cols(nullptr),
-      vals(nullptr),
-      nnz(-1),
-      n_rows(-1),
-      n_cols(-1),
-      device(device),
-      owner(true) {}
+    * @param alloc: the device allocator to use for the underlying buffers
+    * @param stream: CUDA stream to use
+    */
+  COO(std::shared_ptr<deviceAllocator> alloc, cudaStream_t stream)
+    : rows(alloc, stream, 0),
+      cols(alloc, stream, 0),
+      vals(alloc, stream, 0),
+      nnz(0),
+      n_rows(0),
+      n_cols(0) {}
 
   /**
-        * @param rows: coo rows array
-        * @param cols: coo cols array
-        * @param vals: coo vals array
-        * @param nnz: size of the rows/cols/vals arrays
-        * @param n_rows: number of rows in the dense matrix
-        * @param n_cols: number of cols in the dense matrix
-        * @param device: are the underlying arrays on device?
-        */
-  COO(int *rows, int *cols, T *vals, int nnz, int n_rows = -1, int n_cols = -1,
-      bool device = true)
+    * @param rows: coo rows array
+    * @param cols: coo cols array
+    * @param vals: coo vals array
+    * @param nnz: size of the rows/cols/vals arrays
+    * @param n_rows: number of rows in the dense matrix
+    * @param n_cols: number of cols in the dense matrix
+    */
+  COO(device_buffer<Index_Type> &rows, device_buffer<Index_Type> &cols,
+      device_buffer<T> &vals, Index_Type nnz, Index_Type n_rows = -1,
```

Comment:
similar to CSR ctor, the default here has been kept as -1, whereas the other ctors still seem to be defaulting to 0?